### PR TITLE
fix xgboost build failure introduced by allgather interface

### PR DESCRIPTION
**Pull request by chenqin**
detail change https://github.com/dmlc/rabit/pull/113

### CODE REVIEW

**Comment by chenqin** [comment]
@trivialfis

**Comment by thvasilo** [comment]
Pinging @nateagr too. This changes seems strange to me, could you explain briefly why all gather needs these extra inputs to function?

**Comment by chenqin** [comment]
> Pinging @nateagr too. This changes seems strange to me, could you explain briefly why all gather needs these extra inputs to function?

those extra inputs were used to ensure we can do single worker failure recovery before loadcheckpoint called.

File: include/rabit/internal/rabit-inl.h [hunk 1]
```diff
@@ -196,17 +196,22 @@ inline void Allreduce(DType *sendrecvbuf, size_t count,
...
                      _file, _line, _caller);
 }
-#endif  // C++11
 
 // Performs inplace Allgather
 template<typename DType>
-inline void Allgather(DType *sendrecvbuf, size_t totalSize,
-                      size_t beginIndex, size_t sizeNodeSlice,
-                      size_t sizePrevSlice) {
-  engine::Allgather(sendrecvbuf, totalSize * sizeof(DType), beginIndex * sizeof(DType),
```
Comment (chenqin) [inline include/rabit/internal/rabit-inl.h]: this is rootcause, which point to undeclared function

File: src/c_api.cc [hunk 2]
```diff
@@ -131,33 +131,54 @@ void Allgather(void *sendrecvbuf_,
...
     break;
   case kDouble:
     type_size = sizeof(double);
+    rabit::Allgather(static_cast<double*>(sendrecvbuf_), total_size * type_size,
+      beginIndex * type_size, (beginIndex + size_node_slice) * type_size,
+      size_prev_slice * type_size);
     break;
   default: utils::Error("unknown data_type");
   }
-  engine::Allgather
```
Comment (chenqin) [inline src/c_api.cc]: same as above

File: src/engine_empty.cc [hunk 3]
```diff
@@ -26,17 +26,18 @@ class EmptyEngine : public IEngine {
...
+                         size_t slice_end,
+                         size_t size_prev_slice,
+                         const char* _file,
+                         const int _line,
+                         const char* _caller) {
     utils::Error("EmptyEngine:: Allgather is not supported");
   }
   virtual int GetRingPrevRank(void) const {
     utils::Error("EmptyEngine:: GetRingPrevRank is not supported");
+    return -1;
```
Comment (chenqin) [inline src/engine_empty.cc]: a type fix

**Pull request merged**